commitizen/cmd.py

```diff
@@ -12,6 +12,14 @@ class Command(NamedTuple):
     return_code: int
 
 
+def _try_decode(bytes_):
+    try:
+        return bytes_.decode("utf-8")
+    except UnicodeDecodeError:
+        result = chardet.detect(bytes_)
+        return bytes_.decode(result["encoding"] or "utf-8")
```

Comment:
Is the “or” part needed? Doing `decode(“utf-8”)` here could raise unhandled exception.

Reply:
Well, in case of
```
{'encoding': None, 'confidence': 0.0, 'language': None}
```
the `decode()` function should receive the `"utf-8"` encoding arg. Incidentally, that’s the default encoding too ([docs](https://docs.python.org/3/library/stdtypes.html#bytes.decode)) for all Python versions that commitizen supports ([3.6 and later](https://github.com/commitizen-tools/commitizen/blob/2ff9f155435b487057ce5bd8e32e1ab02fd57c94/pyproject.toml#L48)).

Leaving it there or removing it both work, I’m happy with either way.

Reply:
Should we have a minimum threshold? If the confidence is low, what's the point? I would only return if it makes sense, something like:

```python
    try:
        return bytes_.decode("utf-8")
    except UnicodeDecodeError as e:
        result = chardet.detect(bytes_)
        if result["confidence"] > 0.8:
            return bytes_.decode(result["encoding"] or "utf-8")
        raise e
```

Thoughts?

Reply:
I agree with checking the confidence value, the interesting part will be to pick the threshold here 😉

@Woile do you want to reraise the original [UnicodeDecodeError](https://docs.python.org/3/library/exceptions.html#UnicodeDecodeError) if the confidence is too low, or define a new [custom exception](https://github.com/commitizen-tools/commitizen/blob/master/commitizen/exceptions.py) (e.g. `CharacterSetDecodeError` which carries the misidentified charset)?

There’s also the case where a high confidence can still  fail to decode, which we should consider.

Reply:
> Well, in case of
> 
> ```
> {'encoding': None, 'confidence': 0.0, 'language': None}
> ```
> 
> the `decode()` function should receive the `"utf-8"` encoding arg. 

Since I’m using chardet, I’ve never received such dict as answer, but ok to keep that. 


> > Doing `decode(“utf-8”)` here could raise unhandled exception.
> 
> And any other encoding could also raise an exception, too. Which is why @Woile [mentioned](https://github.com/commitizen-tools/commitizen/issues/544#issuecomment-1203810504) to propagate an error and I agree with that.

I did not see that issue, I agree with that comment.

Reply:
> Since I’m using chardet, I’ve never received such dict as answer, but ok to keep that.

Easy to enjoy:
```python
>>> import chardet
>>> chardet.detect("ö".encode())
{'encoding': 'TIS-620', 'confidence': 0.99, 'language': 'Thai'}
>>> chardet.detect("🤦🏻‍♂️".encode())
{'encoding': None, 'confidence': 0.0, 'language': None}
```

> I did not see that issue, I agree with that comment.

Thanks. I’ll change the code accordingly.

Reply:
@jenstroeger I like the idea of a custom exception. And you are right that it can still fail. So maybe we can skip the threshold directly and return the failure:

```python
try:
     bytes_.decode("utf-8")
except UnicodeDecodeError as e:
    result = chardet.detect(bytes_)
    try:
        return bytes_.decode(result["encoding"])  # here we don't need to try utf-8 again
    except UnicodeDecodeError as e:
        raise CharacterSetDecodeError("Could not decode this char", e)
```

Reply:
Looking for chardet alternatives,  [charset_normalizer](https://github.com/Ousret/charset_normalizer) seems promising (py versions, supported languages and also emoji), I’ll try it doing some checks in next future.


I cannot try it at the moment, but if you can try the same strings that led to your errors and if it works correctly, I support its use instead of chardet.

Reply:
@jenstroeger do you have some time to update the code? otherwise let me know and I can take over 👍🏻

Reply:
I’ll do that tonight or tomorrow, it’s been busy today…

Reply:
I did [this PR](https://github.com/commitizen-tools/commitizen/pull/550) using [charset_normalizer](https://github.com/Ousret/charset_normalizer) instead of chardet , can you try that branch and see if the issue still happens ?